src/Grpc.AspNetCore.Server/GrpcServiceExtensions.cs

```diff
@@ -29,22 +30,59 @@ namespace Microsoft.Extensions.DependencyInjection
     public static class GrpcServicesExtensions
     {
         /// <summary>
-        /// Add a GRPC service implementation.
+        /// Adds service specific options to an <see cref="IGrpcServerBuilder"/>.
+        /// </summary>
+        /// <typeparam name="TService">The service type to configure.</typeparam>
+        /// <param name="grpcBuilder">The <see cref="IGrpcServerBuilder"/>.</param>
+        /// <param name="configure">A callback to configure the service options.</param>
+        /// <returns>The same instance of the <see cref="IGrpcServerBuilder"/> for chaining.</returns>
+        public static IGrpcServerBuilder AddServiceOptions<TService>(this IGrpcServerBuilder grpcBuilder, Action<GrpcServiceOptions<TService>> configure) where TService : class
+        {
+            if (grpcBuilder == null)
+            {
+                throw new ArgumentNullException(nameof(grpcBuilder));
+            }
+
+            grpcBuilder.Services.AddSingleton<IConfigureOptions<GrpcServiceOptions<TService>>, GrpcServiceOptionsSetup<TService>>();
+            grpcBuilder.Services.Configure(configure);
+            return grpcBuilder;
+        }
+
+        /// <summary>
+        /// Adds gRPC services to the specified <see cref="IServiceCollection" />.
         /// </summary>
         /// <param name="services">The <see cref="IServiceCollection"/> for adding services.</param>
-        /// <returns></returns>
-        // TODO(JunTaoLuo): Options?
-        public static IServiceCollection AddGrpc(this IServiceCollection services)
+        /// <returns>An <see cref="IGrpcServerBuilder"/> that can be used to further configure the gRPC services.</returns>
+        public static IGrpcServerBuilder AddGrpc(this IServiceCollection services)
```

Comment:
Why do we need a separate Builder type?

Reply:
https://github.com/grpc/grpc-dotnet/blob/43cd16735b633e6b0d8cdc5837dc48726b87291d/test/Grpc.AspNetCore.Server.Tests/GrpcServicesExtensionsTests.cs#L64-L67

Reply:
Can this not be done through just a ServiceCollection? It looks like IGrpcServerBuilder is just a wrapper around it anyway. Is there more things you want to add to the interface later?

Reply:
If it was on a ServiceCollection then you could call AddServiceOptions without AddGrpc, which doesn't make sense. It is there for intelli-sense. MVC and SignalR do the same thing.

Reply:
Sounds good